nixos/modules/misc/nixpkgs.nix

```diff
@@ -14,32 +14,57 @@ let
     then f x
     else f;
 
-  mergeConfig = lhs_: rhs_:
-    let
-      lhs = optCall lhs_ { inherit pkgs; };
-      rhs = optCall rhs_ { inherit pkgs; };
-    in
-    recursiveUpdate lhs rhs //
-    optionalAttrs (lhs ? packageOverrides) {
-      packageOverrides = pkgs:
-        optCall lhs.packageOverrides pkgs //
-        optCall (attrByPath ["packageOverrides"] ({}) rhs) pkgs;
-    } //
-    optionalAttrs (lhs ? perlPackageOverrides) {
-      perlPackageOverrides = pkgs:
-        optCall lhs.perlPackageOverrides pkgs //
-        optCall (attrByPath ["perlPackageOverrides"] ({}) rhs) pkgs;
-    };
-
-  configType = mkOptionType {
+  mergableConfigType = mkOptionType {
     name = "nixpkgs-config";
     description = "nixpkgs config";
-    check = x:
-      let traceXIfNot = c:
-            if c x then true
-            else lib.traceSeqN 1 x false;
-      in traceXIfNot isConfig;
-    merge = args: fold (def: mergeConfig def.value) {};
+    check = isConfig;
+    merge = loc: defs:
+      let
+        # Ensure that each attribute from nixpkgs.config appears at the
+        # first iteration to make sure that nothing is lost.
+        allValues = fold recursiveUpdate {} defs;
+
+        # Recurses through the `config` attr-set for nixpkgs and ensures that
+        # no conflicting definitions exist and that overrides are resolved properly.
+        mapFun = path: value:
+          if isAttrs value && !(value ? _type) then value
+          else let
+            # Evaluate mkIf/mkMerge with `lib.dischargeProperties` and append the
+            # value to the list of declarations. If a value is `null`, it means that
+            # it was not declared in one of the configs and can be ignored.
+            appendVal = all: value_: next:
+              let
+                # FIXME support `mkMerge [{ snens = mkIf cond val; }]`
+                discharged = dischargeProperties value_;
```

Comment:
As mentioned in the comment, `mk{If,Merge,Before,...}` aren't evaluated inside a `mkMerge` (`mkMerge [ (mkIf cond { /* ... */ } ]`) works fine though). IMHO this would make the code even more complex, but are there any suggestions how fix this? :)